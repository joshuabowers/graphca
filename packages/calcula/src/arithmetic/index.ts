import { add, subtract, Addition, isAddition } from "./addition";
import { 
  multiply, divide, negate, double, Multiplication, isMultiplication 
} from "./multiplication";
import { 
  raise, reciprocal, square, sqrt, Exponentiation, isExponentiation 
} from "./exponentiation";

export { 
  add, subtract, 
  multiply, divide, negate, double,
  raise, reciprocal, square, sqrt
}
export {
  Addition, Multiplication, Exponentiation
}
export {
  isAddition, isMultiplication, isExponentiation
}
