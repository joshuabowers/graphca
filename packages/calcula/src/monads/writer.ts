import { method, multi, Multi, fromMulti, _ } from '@arrows/multimethod'

export interface Operation {
  input: unknown,
  action: string
}

export interface Writer<T> {
  value: T,
  log: Operation[]
}

export type WriterFn<T, U = T> = (value: T) => Writer<U>

export const unit = <T>(value: T): Writer<T> => ({value, log: []})

export const bind = <T, U = T>(writer: Writer<T>, transform: WriterFn<T, U>): Writer<U> => {
  const {value, log} = writer
  const {value: result, log: updates} = transform(value)
  return {value: result, log: [...log, ...updates]}
}

export const pipe = <T>(writer: Writer<T>, ...transforms: WriterFn<T>[]): Writer<T> =>
  transforms.reduce(bind, writer)



export const $kind = Symbol('$kind')

type Kind<T extends string> = {[$kind]: T}
type Form<K extends string, R extends {}> = Kind<K> & R

type Real = Form<'Real', {value: number}> 
type Complex = Form<'Complex', {a: number, b: number}> 
type Boolean = Form<'Boolean', {value: boolean}> 
type Nil = Kind<'Nil'>
type NaN = Form<'NaN', {value: number}>
type Primitive = Real | Complex | Boolean | Nil | NaN

type Variable = {[$kind]: 'Variable', name: string, value: Writer<Node>}

type BinaryFields = {left: Writer<Node>, right: Writer<Node>}
type BinaryForm<T extends string> = Form<T, BinaryFields>

type Addition = BinaryForm<'Addition'>
type Multiplication = BinaryForm<'Multiplication'>
type BinaryOperators = Addition | Multiplication

type Equality = BinaryForm<'Equality'>
type Inequality = BinaryForm<'Inequality'>
type Inequalities = Equality | Inequality

type Conjunction = BinaryForm<'Conjunction'>
type Connectives = Conjunction

type Binary = BinaryOperators | Inequalities | Connectives

type UnaryFields = {expression: Writer<Node>}
type UnaryForm<T extends string> = Form<T, UnaryFields>

type Absolute = UnaryForm<'Absolute'> 
type Factorial = UnaryForm<'Factorial'>

type Sine = UnaryForm<'Sine'> 
type Cosine = UnaryForm<'Cosine'>
type Trigonometric = Sine | Cosine

type Unary = Absolute | Factorial | Trigonometric

type Node = Primitive | Unary | Binary | Variable | Nil

type GuardFn<T> = (value: unknown) => value is T
type CreateFn<T, U> = (value: T) => U
type CastFn<T, U> = (value: T) => Writer<U>
type WhenFn<T, U, V> = (create: CreateFn<V, U>) => CastFn<T, U>

type Action<T> = [T|Writer<T>, string]
type CaseFn<I> = (input: I) => Action<I>

type CreateCase<T, U, I> = (create: CreateFn<U, T>) => (input: I) => Action<T>

const is = (kind: Kinds) => <T extends Node>(t: Writer<T>) => t?.value?.[$kind] === kind

const isWriter = <T>(obj: unknown): obj is Writer<T> =>
  typeof obj === 'object' && ('value' in (obj ?? {})) && ('log' in (obj ?? {}))

const isUnary = (v: unknown): v is Writer<Unary> => 
  typeof v === 'object' && v! && isWriter(v) 
  && typeof v.value === 'object' && v.value! && 'expression' in v.value

const isBinary = (v: unknown): v is Writer<Binary> =>
  isWriter(v) && typeof v.value === 'object' && v.value! 
  && 'left' in v.value && 'right' in v.value

const areKindEqual = <T extends Node, U extends Node>(
  t: Writer<T>, u: Writer<U>
) => t.value[$kind] === u.value[$kind]

type KindPredicate<T extends Node> = (v: unknown) => v is Writer<T>
type CaseOfPredicate<T extends Node> = (left: T, right: T) => boolean
const caseOf = <T extends Node>(kind: Kinds | KindPredicate<T>) => 
  (fn: CaseOfPredicate<T> | boolean) =>
    method(
      typeof kind === 'string' ? [is(kind), is(kind)] : [kind, kind], 
      (l: Writer<T>, r: Writer<T>) => 
        areKindEqual(l, r) && (typeof fn === 'boolean' ? fn : fn(l.value, r.value)
      )
    )

type EqualsFn<T extends Node> = (left: Writer<T>, right: Writer<T>) => boolean
type DeepEqualsFn = Multi
  & EqualsFn<Real>
  & EqualsFn<Complex>
  & EqualsFn<Boolean>
  & EqualsFn<Nil>
  & EqualsFn<NaN>
  & EqualsFn<Variable>
  & EqualsFn<Unary>
  & EqualsFn<Binary>
  & EqualsFn<Node>

export const deepEquals: DeepEqualsFn = multi(
  caseOf<Real>('Real')((l, r) => l.value === r.value),
  caseOf<Complex>('Complex')((l, r) => l.a === r.a && l.b === r.b),
  caseOf<Boolean>('Boolean')((l, r) => l.value === r.value),
  caseOf<Nil>('Nil')(true),
  caseOf<NaN>('NaN')(false),
  caseOf<Variable>('Variable')((l, r) => l.name === r.name && deepEquals(l.value, r.value)),
  caseOf<Unary>(isUnary)((l, r) => deepEquals(l.expression, r.expression)),
  caseOf<Binary>(isBinary)(
    (l, r) => deepEquals(l.left, r.left) && deepEquals(l.right, r.right)
  ),
  method(false)
)

const isValue = <T extends Node>(expected: Writer<T>) =>
  <U extends Node>(actual: Writer<U>) =>
    deepEquals(expected, actual)

type WalkFn<T extends Node, R extends Node = T> = (t: Writer<T>) => Writer<R>
const deepEqualsAt = <L extends Node, R extends Node>(
  leftWalk: WalkFn<L, Node>,
  rightWalk: WalkFn<R, Node>
) => (l: Writer<L>, r: Writer<R>) => deepEquals(leftWalk(l), rightWalk(r))

type PrimitiveFn<T, U> = Multi
  & CastFn<U, T>
  & CastFn<Writer<Real>, T>
  & CastFn<Writer<Complex>, T>
  & CastFn<Writer<Boolean>, T>

const primitiveMap = <T, U>(create: CreateFn<U, T>) =>
  <I>(fn: CreateCase<T, U, I>) =>
    (writer: Writer<I>) =>
      bind(writer, input => {
        const [value, action] = fn(create)(input)
        return ({
          value: isWriter(value) ? value.value : value,
          log: [...(isWriter(value) ? value.log : []), {input, action}]
        })
      })

const primitive = <T extends Primitive, U>(
  guard: GuardFn<U>,
  create: CreateFn<U, T>
) => {
  const pMap = primitiveMap(create)
  return (
    whenReal: CreateCase<T, U, Real>,
    whenComplex: CreateCase<T, U, Complex>,
    whenBoolean: CreateCase<T, U, Boolean>
  ) => {
    const fn: PrimitiveFn<T, U> = multi(
      (v: Writer<Node>) => v?.value?.[$kind],
      method(guard, (n: U) => unit(create(n))),
      method('Real', pMap(whenReal)),
      method('Complex', pMap(whenComplex)),
      method('Boolean', pMap(whenBoolean))
    )
    return (
      ...methods: (typeof method)[]
    ): typeof fn => methods.length > 0 ? fromMulti(...methods)(fn) : fn
  }  
}

export const nil: Writer<Nil> = unit({[$kind]: 'Nil'})
export const nan: Writer<NaN> = unit({[$kind]: 'NaN', value: NaN})

export const variable = (name: string, value: Writer<Node> = nil): Writer<Variable> => 
  unit({[$kind]: 'Variable', name, value})

const isNumber = (v: unknown): v is number => typeof v === 'number'
const isNumberTuple = (v: unknown): v is [number, number] =>
  Array.isArray(v) && v.length === 2 && isNumber(v[0]) && isNumber(v[1])
const isBoolean = (v: unknown): v is boolean => typeof v === 'boolean'

export const real = primitive<Real, number>(
  isNumber,
  value => ({[$kind]: 'Real', value})
)(
  _create => r => [r, ''],
  create => c => [create(c.a), 'cast to real'],
  create => b => [create(b.value ? 1 : 0), 'cast to real']
)()

export const complex = primitive<Complex, [number, number]>(
  isNumberTuple,
  ([a, b]) => ({[$kind]: 'Complex', a, b})
)(
  create => r => [create([r.value, 0]), 'cast to complex'],
  _create => c => [c, ''],
  create => b => [create([b.value ? 1 : 0, 0]), 'cast to complex']
)()

export const boolean = primitive<Boolean, boolean>(
  isBoolean,
  value => ({[$kind]: 'Boolean', value})
)(
  create => r => [create(r.value !== 0), 'cast to boolean'],
  create => c => [create(c.a !== 0 || c.b !== 0), 'cast to boolean'],
  _create => b => [b, '']
)()

type UnaryFn<T> = Multi
  & CastFn<Writer<Real>, Real>
  & CastFn<Writer<Complex>, Complex>
  & CastFn<Writer<Boolean>, Boolean>
  & CastFn<Writer<Nil|NaN>, NaN>
  & CastFn<Writer<Node>, T>

const unaryMap = <T>(fn: CaseFn<T>) =>
  (writer: Writer<T>) =>
    bind(writer, input => {
      const [value, action] = fn(input)
      return ({
        value: isWriter(value) ? value.value : value,
        log: [...(isWriter(value) ? value.log : []), {input, action}]
      })
    })

const whenNilOrNaN: CaseFn<Nil | NaN> = _input => [nan.value, 'not a number']

const unary = <T extends Unary>(
  kind: Kinds
) => {
  const create = (expression: Writer<Node>): Action<T> => [
    ({[$kind]: kind, expression}) as T,
    kind.toLocaleLowerCase()
  ]
  return (
    whenReal: CaseFn<Real>,
    whenComplex: CaseFn<Complex>,
    whenBoolean: CaseFn<Boolean>
  ) => {
    const fn: UnaryFn<T> = multi(
      (v: Writer<Node>) => v?.value?.[$kind],
      method('Real', unaryMap(whenReal)),
      method('Complex', unaryMap(whenComplex)),
      method('Boolean', unaryMap(whenBoolean)),
      method('Nil', unaryMap(whenNilOrNaN)),
      method('NaN', unaryMap(whenNilOrNaN)),
      method(unaryMap<Node>(input => create(unit(input))))
    )
    return (
      ...methods: (typeof method)[]
    ): typeof fn => methods.length > 0 ? fromMulti(...methods)(fn) : fn
  }
}

export const absolute = unary<Absolute>('Absolute')(
  r => [real(Math.abs(r.value)), 'absolute value'],
  c => [complex([Math.hypot(c.a, c.b), 0]), 'absolute value'],
  b => [b, 'absolute value']
)()

export const sin = unary<Sine>('Sine')(
  r => [real(Math.sin(r.value)), 'computed sine'],
  c => [
    complex([
      Math.sin(c.a) * Math.cosh(c.b),
      Math.cos(c.a) * Math.sinh(c.b)
    ]),
    'computed sine'
  ],
  b => [boolean(real(Math.sin(b.value ? 1 : 0))), 'computed sine']
)()

type BinaryCaseFn<L, R = L, T = L> = (l: L, r: R) => Action<T>
type BinaryCastFn<L, R = L, T = L> = (l: Writer<L>, r: Writer<R>) => Writer<T>

type Choose<D, F> = F extends void ? D : F

type BinaryFn<T, R = void> = Multi
  & BinaryCastFn<Real, Real, Choose<Real, R>>
  & BinaryCastFn<Complex, Complex, Choose<Complex, R>>
  & BinaryCastFn<Boolean, Boolean, Choose<Boolean, R>>
  & BinaryCastFn<Real, Complex, Choose<Complex, R>>
  & BinaryCastFn<Complex, Real, Choose<Complex, R>>
  & BinaryCastFn<Real, Boolean, Choose<Real, R>>
  & BinaryCastFn<Boolean, Real, Choose<Real, R>>
  & BinaryCastFn<Complex, Boolean, Choose<Complex, R>>
  & BinaryCastFn<Boolean, Complex, Choose<Complex, R>>
  & BinaryCastFn<Nil|NaN, Node, NaN>
  & BinaryCastFn<Node, Nil|NaN, NaN>
  & BinaryCastFn<Node, Node, T>

const binaryMap = <L, R, T>(fn: BinaryCaseFn<L, R, T>) =>
  (l: Writer<L>, r: Writer<R>) => 
    bind(l, x => {
      return bind(r, y => {
        const [value, action] = fn(x, y)
        return ({
          value: isWriter(value) ? value.value : value,
          log: [
            ...(isWriter(value) ? value.log : []), 
            {input: [x, y], action}
          ]
        })
      })
    })

const apply = <T, U>(fn: BinaryFn<T, U>) =>
  <L extends Node, R extends Node>(
    changeLeft: CastFn<Writer<L>, L|R>, 
    changeRight: CastFn<Writer<R>, L|R>
  ) =>
    (l: Writer<L>, r: Writer<R>) =>
      fn(changeLeft(l), changeRight(r))

const identity = <T>(t: T) => t
const leftChild = <T extends Binary>(t: Writer<T>) => t.value.left
const rightChild = <T extends Binary>(t: Writer<T>) => t.value.right

type Kinds = Node[typeof $kind]

const notAny = (...args: Kinds[]) => <T extends Node>(t: Writer<T>) => t.value && !args.includes(t.value[$kind])
const any = (...args: Kinds[]) => <T extends Node>(t: Writer<T>) => t.value && args.includes(t.value[$kind])

type UnaryPredicate<L> = (t: Writer<L>) => boolean 
type BinaryPredicate<L, R> = (l: Writer<L>, r: Writer<R>) => boolean
type toKind<T> = T extends {[$kind]: Kinds} ? T[typeof $kind] : never
type Test<T> = UnaryPredicate<T> | toKind<T> | Writer<T> | typeof _
type CorrespondingFn<L, R> = (l: Writer<L>, r: Writer<R>) => Action<Node>

const when = <L extends Node, R extends Node>(
  predicate: Test<L> | [Test<L>, Test<R>] | BinaryPredicate<L, R>, 
  fn: CorrespondingFn<L, R>
) =>
  method(predicate, (l: Writer<L>, r: Writer<R>) => {
    const [result, action] = fn(l, r)
    return ({
      value: isWriter(result) ? result.value : result,
      log: [
        ...l.log, ...r.log,
        {input: [l.value, r.value], action}, 
        ...(isWriter(result) ? result.log : [])
      ]
    })
  })

const binary = <T extends Binary, R = void>(
  kind: Kinds
) => {
  type Result<U extends Node> = R extends void ? U : R
  const create = (left: Writer<Node>, right: Writer<Node>): Action<T> => [({
    [$kind]: kind, left, right
  }) as T, kind.toLocaleLowerCase()]
  return (
    whenReal: BinaryCaseFn<Real, Real, Result<Real>>,
    whenComplex: BinaryCaseFn<Complex, Complex, Result<Complex>>,
    whenBoolean: BinaryCaseFn<Boolean, Boolean, Result<Boolean>>
  ) => {
    let fn: BinaryFn<T, R> = multi(
      method([is('Real'), is('Real')], binaryMap(whenReal)),
      method([is('Complex'), is('Complex')], binaryMap(whenComplex)),
      method([is('Boolean'), is('Boolean')], binaryMap(whenBoolean)),
      when<Nil|NaN, Node>([any('Nil', 'NaN'), _], (_l, _r) => [nan, 'not a number']),
      when<Node, Nil|NaN>([_, any('Nil', 'NaN')], (_l, _r) => [nan, 'not a number']),
      method(binaryMap<Node, Node, T>((l, r) => create(unit(l), unit(r))))
    )
    return (...methods: (typeof method)[]) => {
      fn = fromMulti(
        ...methods,
        method([is('Real'), is('Complex')], apply(fn)(complex, identity)),
        method([is('Complex'), is('Real')], apply(fn)(identity, complex)),
        method([is('Real'), is('Boolean')], apply(fn)(identity, real)),
        method([is('Boolean'), is('Real')], apply(fn)(real, identity)),
        method([is('Complex'), is('Boolean')], apply(fn)(identity, complex)),
        method([is('Boolean'), is('Complex')], apply(fn)(complex, identity))
      )(fn) as typeof fn
      return fn
    }
  }
}

export const partialLeft = <T extends Node, R = void>(fn: BinaryFn<T, R>) =>
  (left: Writer<Node>): UnaryFn<T> =>
    multi(method((right: Writer<Node>) => fn(left, right)))

export const partialRight = <T extends Node, R = void>(fn: BinaryFn<T, R>) =>
  (right: Writer<Node>): UnaryFn<T> =>
    multi(method((left: Writer<Node>) => fn(left, right)))

type MapFn<T extends Node, R extends Node = Node> = (t: Writer<T>) => Writer<R>
export const binaryFrom = <T extends Node, R = void>(fn: BinaryFn<T, R>) =>
  (leftMap: MapFn<Node> | undefined, rightMap: MapFn<Node> | undefined) =>
    multi(
      method(
        (l: Writer<Node>, r: Writer<Node>) => fn(
          leftMap?.(l) ?? l, rightMap?.(r) ?? r
        )
      )
    ) as typeof fn

type AdditionWithPrimitive = Addition & {right: Primitive}

const isPrimitive = any('Real', 'Complex', 'Boolean')

export const add = binary<Addition>('Addition')(
  (l, r) => [real(l.value + r.value), 'real addition'],
  (l, r) => [complex([l.a + r.a, l.b + r.b]), 'complex addition'],
  (l, r) => [
    boolean((l.value || r.value) && !(l.value && r.value)), 
    'boolean addition'
  ]
)(
  when([
    any('Real', 'Complex', 'Boolean'), notAny('Real', 'Complex', 'Boolean')
  ], (l, r) => [add(r, l), 're-order operands']),
  when<Node, Real>([_, isValue(real(0))], (l, _r) => [l, 'additive identity']),
  when<AdditionWithPrimitive, Primitive>(
    (l, r) => is('Addition')(l) 
      && isPrimitive(l.value.right) 
      && isPrimitive(r), 
    (l, r) => [
      add(l.value.left, add(l.value.right, r)), 
      'combine primitives across nesting levels'
    ]),
  when(deepEquals, (l, _r) => [
    multiply(real(2), l), 'equivalence: replaced with double'
  ]),
  when<Addition, Node>(deepEqualsAt(leftChild, identity), (l, _r) => [
    add(multiply(real(2), l.value.left), l.value.right),
    'combined like terms'
  ])
)

export const subtract = binaryFrom(add)(undefined, r => multiply(real(-1), r))

export const multiply = binary<Multiplication>('Multiplication')(
  (l, r) => [real(l.value * r.value), 'real multiplication'],
  (l, r) => [complex([0, 0]), 'complex multiplication'],
  (l, r) => [boolean(false), 'boolean multiplication']
)()

export const negate = partialLeft(multiply)(real(-1))
export const double = partialLeft(multiply)(real(2))

export const equals = binary<Equality, Boolean>('Equality')(
  (l, r) => [boolean(l.value === r.value), 'real equality'],
  (l, r) => [boolean(l.a === r.a && l.b === r.b), 'complex equality'],
  (l, r) => [boolean(l.value === r.value), 'boolean equality']
)(
  when<Unary, Unary>(
    [isUnary, isUnary], 
    (l, r) => [
      equals(l.value.expression, r.value.expression), 
      'unary equality'
    ]
  )
)
