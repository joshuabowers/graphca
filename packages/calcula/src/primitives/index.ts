import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
