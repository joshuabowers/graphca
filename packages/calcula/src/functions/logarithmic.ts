import { Genera, Species, Notation, TreeNode } from "../utility/tree";
import { real, complex, boolean, Complex } from "../primitives";
import { Binary, binary, when, partialLeft } from "../closures/binary";
import { divide } from '../arithmetic'
import { Exponentiation, isExponentiation } from "../arithmetic/exponentiation";
import { deepEquals } from "../utility/deepEquals";

export type Logarithm = Binary<Species.log, Genera.logarithmic>

const lnComplex = (c: Complex) => complex(
  Math.log(Math.hypot(c.raw.a, c.raw.b)),
  Math.atan2(c.raw.b, c.raw.a)
)

export const [log, isLogarithm, $log] = binary<Logarithm>(
  'log', Notation.prefix, Species.log, Genera.logarithmic
)(
  (l, r) => real(Math.log(r.value.raw) / Math.log(l.value.raw)), 
  (l, r) => {
    const n = lnComplex(r.value)
    if(l.value.raw.a === Math.E && l.value.raw.b === 0){ return n }
    return divide(lnComplex(r.value), lnComplex(l.value))
  },
  (l, r) => boolean(l.value.raw || !r.value.raw)
)(
  when<TreeNode, Exponentiation>(
    (l, r) => isExponentiation(r) && deepEquals(l, r.value.left),
    (_l, r) => [r.value.right, 'inverse operation cancellation']
  )
)

const fromLog = partialLeft(log)
export const lb = fromLog(real(2))
export const ln = fromLog(real(Math.E))
export const lg = fromLog(real(10))
