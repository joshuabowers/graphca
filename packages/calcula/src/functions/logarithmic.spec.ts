import { unit } from '../monads/writer'
import { 
  expectCloseTo, expectWriterTreeNode,
  realOps, complexOps, variableOps, logOps, raiseOps
} from '../utility/expectations'
import { Clades, Genera, Species } from '../utility/tree'
import { real, complex } from '../primitives'
import { variable } from '../variable'
import { raise } from '../arithmetic'
import { log, lb, ln, lg, $log } from './logarithmic'
import { Unicode } from '../Unicode'

describe('$log', () => {
  it('generates a Logarithm from two TreeNode parameters', () => {
    expect(
      $log(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.logarithmic, species: Species.log,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('log', () => {
  it('calculates the real log of a real value', () => {
    expectWriterTreeNode(
      log(real(16), real(256)),
      real(2)
    )(
      ...logOps(
        'real logarithm',
        realOps('16'),
        realOps('256'),
        realOps('2')
      )
    )
  })

  it('calculates the complex log of a complex value', () => {
    expectCloseTo(log(complex([0, 1]), complex([1, 2])), complex([0.7048327646991, -0.5122999987267]), 10)
  })

  it('calculates a complex log of a real value', () => {
    expectCloseTo(log(complex([0, 1]), real(10)), complex([0, -1.465871197758]), 10)
  })

  it('calculates a real log of a complex value', () => {
    expectCloseTo(log(real(10), complex([0, 1])), complex([0, 0.6821881769209]), 10)
  })

  it('returns an expression for unbound subtrees', () => {
    expectWriterTreeNode(
      log(variable('x'), variable('y')),
      $log(variable('x'), variable('y'))[0]
    )(
      ...logOps(
        'created logarithm',
        variableOps('x'),
        variableOps('y'),
        []
      )
    )
  })

  it('returns the exponent of a similarly based exponential', () => {
    expectWriterTreeNode(
      log(real(16), raise(real(16), variable('x'))),
      variable('x')
    )(
      ...logOps(
        'inverse operation cancellation',
        realOps('16'),
        raiseOps(
          'created exponentiation',
          realOps('16'),
          variableOps('x'),
          []
        ),
        variableOps('x')
      )
    )
  })
})

describe('lb', () => {
  it('calculates the base 2 logarithm of a real', () => {
    expectWriterTreeNode(
      lb(real(1024)),
      real(10)
    )(
      ...logOps(
        'real logarithm',
        realOps('2'),
        realOps('1024'),
        realOps('10')
      )
    )
  })

  it('returns the exponent of a binary exponential', () => {
    expectWriterTreeNode(
      lb(raise(real(2), variable('x'))),
      variable('x')
    )(
      ...logOps(
        'inverse operation cancellation',
        realOps('2'),
        raiseOps(
          'created exponentiation',
          realOps('2'),
          variableOps('x'),
          []
        ),
        variableOps('x')
      )
    )
  })
})

describe('ln', () => {
  it('calculates the base e logarithm of complex 0', () => {
    expectWriterTreeNode(
      ln(complex([0, 0])),
      complex([-Infinity, 0])
    )(
      ...logOps(
        'complex logarithm',
        [
          ...realOps(Unicode.e),
          [`${Unicode.e}+0${Unicode.i}`, 'cast to Complex from Real']
        ],
        complexOps('0','0'),
        complexOps(`-${Unicode.infinity}`, '0')
      )
    )
  })

  it('calculates the natural logarithm of a complex', () => {
    expectCloseTo(ln(complex([3.79890743995, 2.1117859405])), complex([1.4693517444, 0.50735630322]), 10)
  })

  it('returns the exponent of a natural exponential', () => {
    expectWriterTreeNode(
      ln(raise(real(Math.E), variable('x'))),
      variable('x')
    )(
      ...logOps(
        'inverse operation cancellation',
        realOps(Unicode.e),
        raiseOps(
          'created exponentiation',
          realOps(Unicode.e),
          variableOps('x'),
          []
        ),
        variableOps('x')
      )
    )
  })
})

describe('lg', () => {
  it('calculates the base 10 logarithm of a real', () => {
    expectCloseTo(lg(real(1000)), real(3), 10)
  })

  it('returns the exponent of a common exponential', () => {
    expectWriterTreeNode(
      lg(raise(real(10), variable('x'))),
      variable('x')
    )(
      ...logOps(
        'inverse operation cancellation',
        realOps('10'),
        raiseOps(
          'created exponentiation',
          realOps('10'),
          variableOps('x'),
          []
        ),
        variableOps('x')
      )
    )
  })
})
