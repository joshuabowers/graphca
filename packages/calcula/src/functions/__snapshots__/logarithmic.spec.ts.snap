// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lb calculates the base 2 logarithm of a real 1`] = `
[
  [
    "log(2,1024)",
    "identified logarithm",
  ],
  [
    "log([2],1024)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "log({2},[1024])",
    "processed left operand; processing right operand",
  ],
  [
    "1024",
    "created real",
  ],
  [
    "log(2,{1024})",
    "processed right operand",
  ],
  [
    "log(2,1024)",
    "real logarithm",
  ],
  [
    "10",
    "created real",
  ],
]
`;

exports[`lb returns the exponent of a binary exponential 1`] = `
[
  [
    "log(2,(2^x))",
    "identified logarithm",
  ],
  [
    "log([2],(2^x))",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "log({2},[(2^x)])",
    "processed left operand; processing right operand",
  ],
  [
    "(2^x)",
    "identified exponentiation",
  ],
  [
    "([2]^x)",
    "processing left operand",
  ],
  [
    "2",
    "created real",
  ],
  [
    "({2}^[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(2^{x})",
    "processed right operand",
  ],
  [
    "(2^x)",
    "created exponentiation",
  ],
  [
    "log(2,{(2^x)})",
    "processed right operand",
  ],
  [
    "log(2,(2^x))",
    "inverse operation cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`lg returns the exponent of a common exponential 1`] = `
[
  [
    "log(10,(10^x))",
    "identified logarithm",
  ],
  [
    "log([10],(10^x))",
    "processing left operand",
  ],
  [
    "10",
    "created real",
  ],
  [
    "log({10},[(10^x)])",
    "processed left operand; processing right operand",
  ],
  [
    "(10^x)",
    "identified exponentiation",
  ],
  [
    "([10]^x)",
    "processing left operand",
  ],
  [
    "10",
    "created real",
  ],
  [
    "({10}^[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(10^{x})",
    "processed right operand",
  ],
  [
    "(10^x)",
    "created exponentiation",
  ],
  [
    "log(10,{(10^x)})",
    "processed right operand",
  ],
  [
    "log(10,(10^x))",
    "inverse operation cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`ln calculates the base e logarithm of complex 0 1`] = `
[
  [
    "log(𝒆,0+0𝒊)",
    "identified logarithm",
  ],
  [
    "log([𝒆],0+0𝒊)",
    "processing left operand",
  ],
  [
    "𝒆",
    "created real",
  ],
  [
    "𝒆+0𝒊",
    "cast to Complex from Real",
  ],
  [
    "log({𝒆+0𝒊},[0+0𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "0+0𝒊",
    "created complex",
  ],
  [
    "log(𝒆+0𝒊,{0+0𝒊})",
    "processed right operand",
  ],
  [
    "log(𝒆+0𝒊,0+0𝒊)",
    "complex logarithm",
  ],
  [
    "-∞+0𝒊",
    "created complex",
  ],
]
`;

exports[`ln returns the exponent of a natural exponential 1`] = `
[
  [
    "log(𝒆,(𝒆^x))",
    "identified logarithm",
  ],
  [
    "log([𝒆],(𝒆^x))",
    "processing left operand",
  ],
  [
    "𝒆",
    "created real",
  ],
  [
    "log({𝒆},[(𝒆^x)])",
    "processed left operand; processing right operand",
  ],
  [
    "(𝒆^x)",
    "identified exponentiation",
  ],
  [
    "([𝒆]^x)",
    "processing left operand",
  ],
  [
    "𝒆",
    "created real",
  ],
  [
    "({𝒆}^[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(𝒆^{x})",
    "processed right operand",
  ],
  [
    "(𝒆^x)",
    "created exponentiation",
  ],
  [
    "log(𝒆,{(𝒆^x)})",
    "processed right operand",
  ],
  [
    "log(𝒆,(𝒆^x))",
    "inverse operation cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`log calculates the real log of a real value 1`] = `
[
  [
    "log(16,256)",
    "identified logarithm",
  ],
  [
    "log([16],256)",
    "processing left operand",
  ],
  [
    "16",
    "created real",
  ],
  [
    "log({16},[256])",
    "processed left operand; processing right operand",
  ],
  [
    "256",
    "created real",
  ],
  [
    "log(16,{256})",
    "processed right operand",
  ],
  [
    "log(16,256)",
    "real logarithm",
  ],
  [
    "2",
    "created real",
  ],
]
`;

exports[`log returns an expression for unbound subtrees 1`] = `
[
  [
    "log(x,y)",
    "identified logarithm",
  ],
  [
    "log([x],y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "log({x},[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "log(x,{y})",
    "processed right operand",
  ],
  [
    "log(x,y)",
    "created logarithm",
  ],
]
`;

exports[`log returns the exponent of a similarly based exponential 1`] = `
[
  [
    "log(16,(16^x))",
    "identified logarithm",
  ],
  [
    "log([16],(16^x))",
    "processing left operand",
  ],
  [
    "16",
    "created real",
  ],
  [
    "log({16},[(16^x)])",
    "processed left operand; processing right operand",
  ],
  [
    "(16^x)",
    "identified exponentiation",
  ],
  [
    "([16]^x)",
    "processing left operand",
  ],
  [
    "16",
    "created real",
  ],
  [
    "({16}^[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(16^{x})",
    "processed right operand",
  ],
  [
    "(16^x)",
    "created exponentiation",
  ],
  [
    "log(16,{(16^x)})",
    "processed right operand",
  ],
  [
    "log(16,(16^x))",
    "inverse operation cancellation",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;
